README.md
[![Build Status](https://github.com/kurtlawrence/papyrus/workflows/Rust%20Tests/badge.svg)](https://github.com/kurtlawrence/papyrus/actions)
[![Latest Version](https://img.shields.io/crates/v/papyrus.svg)](https://crates.io/crates/papyrus)
[![Rust Documentation](https://img.shields.io/badge/api-rustdoc-blue.svg)](https://docs.rs/papyrus)
[![codecov](https://codecov.io/gh/kurtlawrence/papyrus/branch/master/graph/badge.svg)](https://codecov.io/gh/kurtlawrence/papyrus)
[![Rustc Version 1.42+](https://img.shields.io/badge/rustc-1.42+-blue.svg)](https://blog.rust-lang.org/2020/03/12/Rust-1.42.html)

## _Papyrus_ - A rust REPL and script running tool.

> This project is no longer maintained.

See the [rs docs](https://docs.rs/papyrus/) and the
[guide](https://kurtlawrence.github.io/papyrus/).
Look at progress and contribute on [github.](https://github.com/kurtlawrence/papyrus)

```sh
papyrus=> 2+2
papyrus [out0]: 4
```

## Overview
Papyrus creates a Rust REPL in your terminal. Code can be typed in, line by line with feedback on
the evaluation, or code can be injected via stdin handles. 
Each code snippet is evaluated on an expression based system, so terminating with a semi-colon
requires more input.

### Example
```sh
[lib] papyrus=> 2+2
papyrus [out0]: 4
[lib] papyrus=> println!("Hello, world!");
[lib] papyrus.> out0 * out0
Hello, world!
papyrus [out1]: 16
[lib] papyrus=> :help
help -- prints the help messages
cancel | c -- returns to the root class
exit -- sends the exit signal to end the interactive loop
Classes:
  edit -- Edit previous input
  mod -- Handle modules
Actions:
  mut -- Begin a mutable block of code
[lib] papyrus=> :exit
[lib] papyrus=> Thanks for using papyrus!
```

## Installation
Papyrus can be installed from `crates.io` or building from source on github.
The default installation feature set requires a `nightly` toolchain, but `stable` can be used with
fewer features enabled.

To install with all features:
```sh
rustup toolchain add nightly
cargo +nightly install papyrus
```

To install on stable without racer completion:
```sh
cargo +stable install papyrus --no-default-features --features="format,runnable"
```

## Requirements

### Features
Papyrus has features sets:
- _format_: format code snippets using `rustfmt`
- _racer-completion_: enable code completion using [`racer`](https://github.com/racer-rust/racer).
    **Requires a nightly compiler**
- _runnable_: papyrus can be _run_, without needing to manually handle repl states and output

All features are enabled by default.

### Cargo
Papyrus leverages installed binaries of both `cargo` and `rustc`. This requirement may lift in the
future but for now, any user wanting to use Papyrus will need an installation of Rust.



CHANGELOG.md
# Changelog

## 0.17.0
- Path to examples in README fixed
- REPL `kserd` has `format` feature enabled
- Libraries are `Box`ed to avoid possible segmentation faults
- Added an on exit function to `RunCallbacks`.
- Terminal interface now has history that can be retrieved using Up and Down arrows
- `:static-files` supports glob patterns.
- Added `:mod clear` that clears previous REPL inputs

## 0.17.1
- Fix compiling bug with latest `crossbeam_channel` release

## 0.17.2
- Fix compiling bug with latest `syn` release

## 0.16.0
- Increase `libloading` dependency to `0.6`.
- Increase `kserd` dependency to `0.4`.
- Move MSRV to `1.42` to support `libloading`.

## 0.15.0
- Added ability to have static files, through API and REPL interface.

## 0.14.0
- Fix completion overwriting lines before bug #59
- Add `persistent_module_code` on the `LinkingConfiguration`, this is used to solve bug #57
- **Breaking Change:** `AppDataAction` now expects closure that has access to `&mut ReplData` - #60
- **Breaking Change:** Add functionality to inject callbacks when running a REPL - #58
- Output `.dylib` on MacOS - #66
- Update to `kserd` 0.3
- **Breaking Change:** Renaming of `to_*` to `into_*` in various places.
- **Breaking Change:** Removed `SourceCode::new` in-lieu of `Default` impl.
- **Breaking Change:** `format()` no _does not replace new lines with spaces_.
- Rework of interface. Increased testability and robustness.
- **Breaking Change:** Increase MSRV to `1.39`

### 0.14.1
- Fix compilation issue with `backtrace` dependency

## 0.13.0
- Restructure of repository
- Fix unintended indenting on unix when using `println!` macro
- Fix drawing issues when input would overflow into other lines
- Libraries can now be cached and not dropped once evaluation is finished.
- Removed ability to redirect output of evaluation.
  - This functionality is broken and requires 
    a. More testing
    b. Better use case
- Fixed the high cpu usage regression
- Increase requirement of MSRV to 1.36

### 0.13.1
- Up the event waiting duration from 1 millisecond to 5

## 0.12.0
- Handle `Item::Use` when parsing input.
- Use of `backtrace` crate to dump complete backtrace with panic handling.
- Update `crossterm` to 0.13.
- Remove `rustfmt-nightly` from dependencies, using `rustfmt` binary to format code
  - This was done as building on nightly crashed too often.
- Handle inner attribute syntax such as `#![feature(test)]`.
- Changed `Item` from `String` to `(String, bool)`.
- Remove `mortal` dependency, moving to `crossterm` for input. Fixes erratic typing input and not
    showing on windows.
- Bug fixes

### 0.12.1
- Handle `Item::Macro`.
- Updated linking documentation.
- Code improvements
- Terminal interface input events are buffered so multiline stdin is handled correctly.
- Update `crossterm` dependency to 0.14.

## 0.11.0

- docs: Add `cmds` module documentation.
- docs: Add `complete` module documentation.
- Move `code` and `linking` module to root level.

## 0.10.0

- Removed `Repl.run_with_completion()`, use `Repl.run()` with `racer-completion` feature instead.
- docs: Update output module documentation.
- New `runnable` interface. Foundational change to allow for more flexiblity and growth.
- Switch command prefix `.` to `:`.
- Add `Repl.run_async()` function that accepts an `Arc<Mutex<D>>` for data.
- Change `eval_async` to accept `Arc<Mutex<D>>`.

## 0.9.0

- mod paths now complete
- No longer need a terminal interface, maintains own internal buffer for output
- Fixed bug where using as library would not compile.

## 0.8.1

- Re-export `azul` crate in `prelude`.
- Re-export `linefeed::memory::MemoryTerminal` struct in `prelude`.
- Re-export `cmdtree` items into the `prelude`.
- Added `CommandResult::Empty`
- Updated `cmdtree` to `v0.5` so users can output to a writer.
- Added `CommandResult::ActionOnAppData` so user can take actions on app data
- `CommandResult` actions pass through writer.
- async repls require `RwLock` rather than `Mutex`.
- Added module support

## 0.8.0

- The macros have been changed. Now `repl!` and `repl_with_term!` are used to create a `Repl` instance, with access to the public `.data` field to customise the repl data.
- there is no more variants for borrowing patterns, only the specified data type
- Added a `push_input()` function a `Read` repl. This allows for pushing of characters onto the repl without reading from `stdin`.
- `Repl` now takes ownership of `ReplData`. It did not make sense to borrow as you would have to drop `Repl` to make any changes to `ReplData`...
- Added `push_input_str()` - an extension of `push_input()`
- Added `eval_async` functions for `Repl`s in the `Evaluate` stage
- `widgets::pad` is now released!

## 0.7.0

Major change in api as the library is shifted towards a new repl direction.

- `Repl`s are now state machines
- Cannot construct a `Repl` using a file
- Cannot `evaluate` off a `Repl` (this has been altered to work with state machine)
- `Command` now requires to return a repl in print state (`Result<Repl<Print>, ()>`)
- `Repl`s use `ReplData` which is passed between states and instantiated outside the repl machine
- Documentation is not up to date, and will not be until some features stablise
- `Repl`s can now link external crates and data.
- Removed context menu functionality
- Removed repl file loading

## 0.6.1

- Papyrus now formats the source code written to file.
- Help messages have been colourized for clarity.

## 0.6.0

`papyrus` now works with stable rust! 🎉

```sh
rustup default stable
cargo install papyrus
```

## 0.5.2

- Added the `.cancel` and `.c` commands which allow users to cancel out of the current input. This lets you exit more input loops if a leading closing bracket was defined.

## 0.5.1

- Turned off colouring for Windows, not yet working as intended.

## 0.5.0

- Added a version query to `crates.io`. When the repl is run interactively it will check the version number and print if there is a later version available.

src/lib.rs
//! [![Build Status](https://github.com/kurtlawrence/papyrus/workflows/Rust%20Tests/badge.svg)](https://github.com/kurtlawrence/papyrus/actions)
//! [![Latest Version](https://img.shields.io/crates/v/papyrus.svg)](https://crates.io/crates/papyrus)
//! [![Rust Documentation](https://img.shields.io/badge/api-rustdoc-blue.svg)](https://docs.rs/papyrus)
//! [![codecov](https://codecov.io/gh/kurtlawrence/papyrus/branch/master/graph/badge.svg)](https://codecov.io/gh/kurtlawrence/papyrus)
//! [![Rustc Version 1.42+](https://img.shields.io/badge/rustc-1.42+-blue.svg)](https://blog.rust-lang.org/2020/03/12/Rust-1.42.html)
//!
//! ## _Papyrus_ - A rust REPL and script running tool.
//!
//! See the [rs docs](https://docs.rs/papyrus/) and the
//! [guide](https://kurtlawrence.github.io/papyrus/).
//! Look at progress and contribute on [github.](https://github.com/kurtlawrence/papyrus)
//!
//! ```sh
//! papyrus=> 2+2
//! papyrus [out0]: 4
//! ```
//!
//! Papyrus is in active development, see [changelog](https://github.com/kurtlawrence/papyrus) for updates.
//!
//! ## Overview
//! Papyrus creates a Rust REPL in your terminal. Code can be typed in, line by line with feedback on
//! the evaluation, or code can be injected via stdin handles.
//! Each code snippet is evaluated on an expression based system, so terminating with a semi-colon
//! requires more input.
//!
//! ### Example
//! ```sh
//! [lib] papyrus=> 2+2
//! papyrus [out0]: 4
//! [lib] papyrus=> println!("Hello, world!");
//! [lib] papyrus.> out0 * out0
//! Hello, world!
//! papyrus [out1]: 16
//! [lib] papyrus=> :help
//! help -- prints the help messages
//! cancel | c -- returns to the root class
//! exit -- sends the exit signal to end the interactive loop
//! Classes:
//!   edit -- Edit previous input
//!   mod -- Handle modules
//! Actions:
//!   mut -- Begin a mutable block of code
//! [lib] papyrus=> :exit
//! [lib] papyrus=> Thanks for using papyrus!
//! ```
//!
//! ## Installation
//! Papyrus can be installed from `crates.io` or building from source on github.
//! The default installation feature set requires a `nightly` toolchain, but `stable` can be used with
//! fewer features enabled.
//!
//! To install with all features:
//! ```sh
//! rustup toolchain add nightly
//! cargo +nightly install papyrus
//! ```
//!
//! To install on stable without racer completion:
//! ```sh
//! cargo +stable install papyrus --no-default-features --features="format,runnable"
//! ```
//!
//! ## Requirements
//!
//! ### Features
//! Papyrus has features sets:
//! - _format_: format code snippets using `rustfmt`
//! - _racer-completion_: enable code completion using [`racer`](https://github.com/racer-rust/racer).
//!     **Requires a nightly compiler**
//! - _runnable_: papyrus can be _run_, without needing to manually handle repl states and output
//!
//! All features are enabled by default.
//!
//! ### Cargo
//! Papyrus leverages installed binaries of both `cargo` and `rustc`. This requirement may lift in the
//! future but for now, any user wanting to use Papyrus will need an installation of Rust.

#![warn(missing_docs)]
#![deny(intra_doc_link_resolution_failure)]

#[macro_use]
extern crate log;

#[cfg(feature = "runnable")]
#[macro_use]
extern crate crossterm;

/// Build a repl instance with the default terminal.
/// If a type is specfied (ie `repl!(String)`) then the repl will be bounded to use
/// that data type. Otherwise the default `()` will be used.
#[macro_export]
macro_rules! repl {
    // Default Term, with type
    ($type:ty) => {{
        use papyrus;
        let mut r: papyrus::repl::Repl<_, $type> = papyrus::repl::Repl::default();
        r.data = unsafe { r.data.set_data_type(&format!("{}", stringify!($type))) };
        r
    }};

    // No data
    () => {{
        use papyrus;
        let r: papyrus::repl::Repl<_, ()> = papyrus::repl::Repl::default();
        r
    }};
}

#[cfg(debug_assertions)]
#[allow(unused_macros)]
macro_rules! dbg_to_file {
    ($val:expr) => {{
        use std::io::Write;
        let mut file = std::fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open("_dbg_to_file_output")
            .expect("failed to create/open _dbg_to_file_output file");
        writeln!(
            file,
            "[{}:{}] {} = {:?}",
            file!(),
            line!(),
            stringify!($val),
            $val
        )
        .unwrap();
    }};
}

/// The prefix to access commands.
const CMD_PREFIX: &str = ":";

pub mod cmds;
pub mod code;
pub mod compile;
pub mod complete;
/// Format rust code snippets using `rustfmt`.
///
/// Requires the _format_ feature.
#[cfg(feature = "format")]
pub mod fmt;
/// Parsing of input.
pub mod input;
pub mod linking;
pub mod output;
pub mod repl;

/// Running the repl. Requires `runnable` feature.
#[cfg(feature = "runnable")]
pub mod run;

/// Re-exports of most common types and modules.
pub mod prelude {
    pub use crate::repl::{self, ReadResult, Repl, ReplData, Signal};
}

pub use ::cmdtree;
#[cfg(feature = "racer-completion")]
pub use ::racer;
